jsapp/js/designSystem/buttonDemo.tsx

```diff
@@ -0,0 +1,265 @@
+import React from 'react'
+import Select from 'react-select'
+import bem from 'js/bem';
+import Checkbox from 'js/components/common/checkbox'
+import TextBox from 'js/components/common/textBox'
+import Radio from 'js/components/common/radio'
+import Button, {ButtonType, ButtonColor, ButtonSize} from 'js/components/common/button'
+import {IconNames} from 'jsapp/fonts/k-icons'
+
+const buttonTypes: ButtonType[] = ['bare', 'frame', 'full']
+const buttonColors: ButtonColor[] = ['blue', 'red', 'storm']
+const buttonSizes: ButtonSize[] = ['s', 'm', 'l']
+
+const iconNamesOptions: IconNameOption[] = []
+for (let iconName in IconNames) {
+  // Hacking TypeScript: https://stackoverflow.com/a/51281023/2311247
+  const typedIconName: IconNames = IconNames[iconName as keyof typeof IconNames]
+  iconNamesOptions.push({
+    label: typedIconName,
+    value: typedIconName
+  })
+}
+
+type ButtonDemoState = {
+  demoType: ButtonType
+  demoColor: ButtonColor
+  demoSize: ButtonSize
+  demoStartIcon: IconNameOption | null
+  demoEndIcon: IconNameOption | null
+  demoLabel: string
+  demoTooltip: string
+  demoIsDisabled: boolean
+  demoIsPending: boolean
+  demoIsFullWidth: boolean
+}
+
+type IconNameOption = {
+  value: IconNames,
+  label: IconNames
+}
+
+const defaultLabel = 'click me'
+const defaultIcon: IconNameOption = {
+  label: IconNames.trash,
+  value: IconNames.trash
+}
+
+export default class ButtonDemo extends React.Component<{}, ButtonDemoState> {
+  constructor(props: {}) {
+    super(props)
+    this.state = {
+      demoType: 'full',
+      demoColor: 'blue',
+      demoSize: 'l',
+      demoStartIcon: null,
+      demoEndIcon: null,
+      demoLabel: defaultLabel,
+      demoTooltip: '',
+      demoIsDisabled: false,
+      demoIsPending: false,
+      demoIsFullWidth: false,
+    }
+  }
+
+  onTypeChange({}: any, newType: ButtonType) {
+    this.setState({demoType: newType})
+  }
+
+  onColorChange({}: any, newColor: ButtonColor) {
+    this.setState({demoColor: newColor})
+  }
+
+  onSizeChange({}: any, newSize: ButtonSize) {
+    this.setState({demoSize: newSize})
+  }
+
+  onStartIconChange(newStartIcon: IconNameOption | null) {
+    this.setState({
+      demoStartIcon: newStartIcon ? newStartIcon : null,
+      // Only one of icons is allowed.
+      demoEndIcon: newStartIcon ? null : this.state.demoEndIcon
+    })
+  }
+
+  onEndIconChange(newEndIcon: IconNameOption | null) {
+    this.setState({
+      demoEndIcon: newEndIcon ? newEndIcon : null,
+      // Only one of icons is allowed.
+      demoStartIcon: newEndIcon ? null : this.state.demoStartIcon
+    })
+  }
+
+  onLabelChange(newLabel: string) {
+    this.setState({
+      demoLabel: newLabel,
+      // If there is no label, icon is required
+      demoStartIcon: newLabel === '' ? defaultIcon : this.state.demoStartIcon
```

Comment:
Not a big deal at all, but if we clear the text box then both icons in start and end show up:
![Screenshot from 2022-01-03 16-05-26](https://user-images.githubusercontent.com/16762675/147980286-f5c7ff2e-c584-4b01-a6fb-154a51eac253.png)
looks like we would just not do this if `demoEndIcon` already exists

Reply:
Oh I see, in the `onLabelChange` function in `buttonDemo.tsx` I was setting a default icon when label was removed, but I was only setting one of the icons, with other remaining what was already there. Fixed that :) Plus for the `button.tsx` I am now displaying the end icon only if start icon wasn't defined.